---
title: "MetaboAnalystR 2.0 Workflow: From Raw Spectra to Biological Insights"
author: "Jasmine Chong, Jeff Xia"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MetaboAnalystR 2.0 Workflow: From Raw Spectra to Biological Insights}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## 1. Introduction

We present MetaboAnalystR 2.0, which aims to address two important gaps left in its previous version. First, raw spectral processing - the previous version offered very limited support for raw spectra processing and peak annotation. Therefore, we have implemented comprehensive support for raw LC-MS spectral data processing including peak picking, peak alignment and peak annotations. Second, we have enhanced support for functional interpretation directly from m/z peaks. In addition to an efficient implementation of the mummichog algorithm (PMID: 23861661), we have added a new method to support pathway activity prediction based on the well-established GSEA algorithm (PMID: 16199517). In this tutorial, we showcase how to utilize MetaboAnalyst 2.0 to perform a comprehensive end-to-end metabolomics data workflow. The dataset consists of a subset of pediatric IBD stool samples obtained from the Integrative Human Microbiome Project Consortium (https://ibdmdb.org/).

## 2. Raw MS Data Preprocessing

Three main wrapper functions have been implemented for metabolomics data processing based on XCMS (PMIDs: 16448051, 19040729, and 20671148; version 3.4.4) and CAMERA (PMID: 22111785; version 1.38.1) including: (i) the *ImportRawMSData* function for reading in raw data files, (ii) the *PerformPeakProfiling* function for peak picking and alignment, and (iii) the *PerformPeakAnnotation* function for peak annotation. These functions are described below in further detail.

The *ImportRawMSData* function reads in raw MS data files and saves it as an OnDiskMSnExp object. The function outputs two plots - the Total Ion Chromatogram (TIC) which provides an overview of the entire spectra, and the Base Peak Chromatogram (BPC) which is a cleaner profile of the spectra based on the most abundant signals. These plots are useful to inform the setting of parameters downstream. For users who wish to view a peak of interest, an Extracted Ion Chromatogram (EIC) can be generated using the *PlotEIC* function.

The *PerformPeakProfiling* function is a wrapper of several XCMS R functions that performs peak detection, alignment, and grouping in a single step. The resulting peaks are outputted as a XCMSnExp object. The function also generates two diagnostic plots including a retention time adjustment map, and a PCA plot showing the overall sample clustering prior to data cleaning and statistical analysis. Users can specify several parameters such as the mass accuracy, peak width, and retention time range using the SetPeakParam function to optimize the peak picking function.

The *PerformPeakAnnotation* function annotates isotope and adduct peaks using the CAMERA R package (PMID: 22111785). It outputs the result as a CSV file (“annotated_peaklist.csv”) and saves the annotated peaks as an xsAnnotate object. Finally, the peak list is formatted to the correct structure for MetaboAnalystR and filtered based upon user’s specifications using the *FormatPeakList* function. This function permits the filtering of adducts (i.e. removal of all adducts except for [M+H]+/[M-H]-) and filtering of isotopes (i.e. removal of all isotopes except for monoisotopic peaks). The goal of filtering peaks is to remove degenerative signals and reduce the file size.

### 1.1 Example data download

Prior to starting this tutorial, please download the 6 example mzML files, containing fecal samples from 3 pediatric IBD patients and 3 healthy controls. These patients are all male, aged 14-16 years old. The data is available for download here: https://www.dropbox.com/sh/o1nqb7ioyo868nt/AADWDKRsCo6tZkWvDcFBGGhJa?dl=0. Ensure that you unzip these files into an empty folder and set your working directory to this new folder. If you have issues downloading/unzipping the zipped folder, try downloading these 6 samples individually.

### 1.2 Preprocessing of example data

We will first import the raw MS data using the *ImportRawMSData* function. Use the list.files() functions to view the order of the files within the folder. As you can see, the first 3 samples are the controls, and the last 3 samples are IBD patients - denoted with a CD at the beginning of their file names. Therefore, within *ImportRawMSData* we set grpA to "Healthy", numA to 3, grpB to "IBD", and numB to 3. The format refers to the format of the resulting TIC and BPC plots (i.e. "png", "jpg"), the dpi refers to the resolution (typically 300 dpi is required for publication quality images) and width refers to the size of the plot. All plots generated in MetaboAnanlystR can be found within your current working directory.

```{r, eval=FALSE}

# Set working directory to 6 clinical samples
setwd("~/Desktop/iHMP/")

# Load the MetaboAnalystR package
library("MetaboAnalystR")

# Import mzML files
rawData <- ImportRawMSData(grpA = "Healthy", numA = 3, grpB = "IBD", numB = 3, 
                            format = "png", dpi = 72, width = 9)

```

Next, we will set the parameters for peak picking using the *SetPeakParam* function. This function sets all the parameters used for downstream pre-processing of user's raw MS data such as the mass error (ppm), signal to noise threshold, and bandwidth. Use ?SetPeakParam for further information of all available parameters. In this case, we set the mass error to 5. Following the setting of parameters, we will use the *PerformPeakProfiling* to perform peak picking, grouping, retention time correction, and alignment. 

```{r, eval=FALSE}
# Set parameters for peak profiling
peakParams <- SetPeakParam(ppm = 5)

# Perform peak profiling
extPeaks <- PerformPeakProfiling(rawData, peakParams, rtPlot = TRUE, pcaPlot = TRUE,
                               format = "png", dpi = 300, width = 9)

```

The *PerformPeakProfiling* function outputs a PCA plot of the detected peaks per sample. From this plot, we can see a relatively good separation between the Healthy and IBD patients. 

Following peak profiling we will perform peak annotation. First, set the parameters for peak annotation using the *SetAnnotationParam* function. In this case, specify that the samples were collected in negative ion mode - "negative", and the allowed variance for the search (for adduct annotation) to 0.005. Next, use the *PerformPeakAnnotation* function to annotate the detected peaks to potential isotopes and  adducts. This will output a csv file named "annotated_peaklist.csv", which contains the entire list of all identified m/z features, their retention time, and their annotations. In this case, we identified ~11000 features. Finally, use the *FormatPeakList* function to format this annotated_peaklist into the correct format to be used by MetaboAnalystR (and the web version). In this function, users can choose to filter the peaks, i.e. remove features that are missing in more than 75% of features per group and filter out all isotopes. The function will output a file named "metaboanalyst_input.csv", which will be used the package to read in the peak table.

```{r, eval=FALSE}
# Set parameters for peak annotation
annParams <- SetAnnotationParam(polarity = "negative", mz_abs_add = 0.005)

# Perform peak annotation
annotPeaks <- PerformPeakAnnotation(extPeaks, annParams)

# Format and filter the peak list for MetaboAnalystR
maPeaks <- FormatPeakList(annotPeaks, annParams, filtIso = TRUE, filtAdducts = FALSE,
                          missPercent = 0.75)

```

## 3. Data Processing and Statistical Analysis

Remaining in the same working directory, we will read in the filtered peak table. Then, we will replace missing values with a small value (half of the minimum positive value within the original data). Next we will filter varibles out non-informative signals, then normalize the samples to their median and perform log transformation. Finally, we will perform t-test analysis to identify differentially enriched features, setting the FDR-adjusted p-value threshold to 0.25.

```{r, eval=FALSE}
# Initialize the mSetObj
mSet<-InitDataObjects("pktable", "stat", FALSE)

# Read in the filtered peak list
mSet<-Read.TextData(mSet, "metaboanalyst_input.csv", "colu", "disc")

# Perform data processing
mSet<-SanityCheckData(mSet)
mSet<-ReplaceMin(mSet);
mSet<-FilterVariable(mSet, "iqr", "F", 25)
mSet<-PreparePrenormData(mSet)
mSet<-Normalization(mSet, "MedianNorm", "LogNorm", "NULL", ratio=FALSE, ratioNum=20)

# Perform t-test
mSet<-Ttests.Anal(mSet, F, 0.25, FALSE, TRUE)

```

## 4. From Raw Peaks to Biological Insights

Even with high mass accuracy afforded by the current high-resolution MS platforms, it is often impossible to uniquely identify a given peak based on its mass alone. To get around this issue, a key concept is to shift the unit of analysis from individual compounds to individual pathways or a group of functionally related compounds (i.e. metabolite sets (PMID: 20457745)). The general assumption is that the collective behavior of a group is more robust against a certain degree of random errors of individuals. The mummichog algorithm is the first implementation of this concept to infer pathway activities from a ranked MS peaks identified in untargeted metabolomics. This algorithm is available in MetaboAnalystR (the *PerformMummichog* function). The original algorithm implements an over-representation analysis (ORA) method to evaluate pathway-level enrichment based on significant features. Users need to specify a pre-defined cutoff based on either t-statistics or fold changes. A complementary approach is the Gene Set Enrichment Analysis (GSEA) method, a widely used method to extract biological meaning from a ranked gene list (PMID: 16199517). Unlike ORA, this method considers the overall ranks without using a significance cutoff. It is able to detect subtle and consistent changes which can be missed from ORA methods. The MetaboAnalystR 2.0 offers GSEA (the PerformGSEA function) using the high-performance fgsea R package (https://www.biorxiv.org/content/10.1101/060012v1).

To perform these analyses, users must upload a table containing three columns - m/z features, p-values, and t-scores or fold-change values. In this case, as we performed t-test analysis on the peak table, we can use the new function *Convert2Mummichog*, which converts the results of the t-test to the proper format for mummichog analysis. The output of this function is a file titled "mummichog_input_XX.txt", with XX replaced by the current date. 

Next, we follow the typical MS Peaks to Pathways workflow, setting the instrument options to "negative" and the p-value cut-off to 0.25. Then, we will perform the original mummichog algorithm, using the *PerformMummichog* function. We will select the hsa_mfn library, and set the permutations to 1000. This may take some time. 

```{r, eval=FALSE}

# Convert the output of the t-test to the proper input for mummichog analysis
mSet<-Convert2Mummichog(mSet)

# Read in the ranked peak list (don't forget to change the file name!)
mSet<-Read.PeakListData(mSet, "mummichog_input_2019-03-05.txt");
mSet<-UpdateMummichogParameters(mSet, "5", "negative", 0.25);
mSet<-SanityCheckMummichogData(mSet)

# Perform original mummichog algorithm
mSet<- PerformMummichog(mSet, "hsa_mfn", "fisher", "gamma", permNum = 1000)

# View the top pathway hits using head(mSet$mummi.resmat)
# > head(mSet$mummi.resmat)
#                                                           Pathway total Hits.total Hits.sig    EASE      FET     Gamma
# Leukotriene metabolism                                               92         46        3 0.23131 0.065701 0.0016689
# C21-steroid hormone biosynthesis and metabolism                     112         86        3 0.51826 0.258530 0.0058712
# Prostaglandin formation from arachidonate                            78         55        2 0.67555 0.314110 0.0126580
# Androgen and estrogen biosynthesis and metabolism                    95         63        2 0.72565 0.375550 0.0165210
# Putative anti-Inflammatory metabolites formation from EPA            27         22        1 1.00000 0.369520 1.0000000
# Vitamin E metabolism                                                 54         31        1 1.00000 0.479100 1.0000000
```

Now we will perform GSEA, using the *PerformGSEA* function. This function is much faster than the original implementation. As you can see, Vitamin E metabolism is consistently identified between the two algorithms. Vitamin E has been shown to be decreased in patients with Crohn's disease (PMID: 22488830) and were shown to have anti-inflammatory properties (PMID: 30237488). Other pathways also have been associated with IBD in the literature. For instance, increased leukotriene production has been demonstrated in IBD (PMIDs: 6319219, 17230539). Finally, we can visualize the comparison of the pathway analysis results of the two algorithms with a scatter plot using the *PlotIntegPaths* function. By adding labels = TRUE to the function, we can generate a plot with all of the dots annotated - this may be cluttered for those with many pathways identified.

```{r, eval=FALSE}
# Perform GSEA 
mSet<- PerformGSEA(mSet, "hsa_mfn", "fisher", "gamma", permNum = 1000)

# View the top pathway hits using head(mSet$mummi.gsea.resmat)
# > head(mSet$mummi.gsea.resmat)
#                                                           Pathway_Total Hits P_val    P_adj NES      
# Putative anti-Inflammatory metabolites formation from EPA "27"          "1"  "0.6516" "1"   "-0.9053"
# Vitamin E metabolism                                      "54"          "1"  "0.9939" "1"   "-0.6143"
# Androgen and estrogen biosynthesis and metabolism         "95"          "2"  "0.8693" "1"   "0.6874" 
# Drug metabolism - cytochrome P450                         "53"          "1"  "0.9942" "1"   "0.3174" 
# Tryptophan metabolism                                     "94"          "1"  "0.9942" "1"   "0.5734" 
# Urea cycle/amino group metabolism                         "85"          "1"  "0.9942" "1"   "0.3641" 

# Plot the integration of the two algorithms
mSet <- PlotIntegPaths(mSet, dpi = 300, width = 10, format = "jpg")

# Plot with all dots annotated
mSet <- PlotIntegPaths(mSet, dpi = 300, width = 10, format = "jpg", labels = TRUE)

```

